src/ucp/core/ucp_worker.c

```diff
@@ -2481,7 +2482,21 @@ static unsigned ucp_worker_keepalive_progress(void *arg)
             continue;
         }
 
-        ep_keepalive_count += ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
+        status = ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext),
+                                     &worker->keepalive.lane_map);
+        if (status == UCS_OK) {
+            ep_keepalive_count++;
+            /* reset lane map due to processed all uct lanes */
+            worker->keepalive.lane_map = (ucp_lane_map_t)-1;
```

Comment:
maybe init to `ucp_ep_config(ep)->key.ep_check_map`? Then no need to recalculate lane_map in `ucp_ep_do_keepalive`

Reply:
such change will make logic much more complicated: for now we reset flag for any next EP on current iteration, but to set specific mask (like same as ep_check_mask) we have to identify next EP which may process keepalive, or we have to store additional state which identify result of previous operation: ok or in-progress... current implementation need just one extra operation - binary ```and``` on lanes iteration, your proposal will require much more operations

Reply:
what if we initialize worker->keepalive.lane_map when stepping to next ep ? and also, instead of "stepping back" here, we always start from the "current position", and step to next ep only after returning from ucp_ep_do_keepalive() with status != UCS_ERR_NO_RESOURCE?
BTW, instead, of checking return status from ucp_ep_do_keepalive(), we could step to next ep iff worker->keepalive.lane_map became 0 (we expect that if KA is not supported per ep, its ucp_ep_config(ep)->key.ep_check_map would be 0 anyway right?)

Reply:
what should be initial value for lane_map?

Reply:
map of of the first ep

Reply:
the problem is in removing EP from worker: in case if we removing current keepalive element we just update iterator to next element in list & reset map to -1 (BWT - I missed to add this), but if we decide to set exact map prior to keepalive progress we will have to search next suitable EP & take map from it - it could be extremely expensive in some cases

Reply:
ok, updated